### PR TITLE
 Combine various ray plugins into one (issue 772)

### LINKED ISSUE #768: gazebo_ros_laser and gazebo_ros_gpu_laser incompatibility
I try to use both gpu_ray and ray sensors at the same time but it always gets me an error which leads to the none of the laser_scans being published. I am using Gazebo 9.1 and ros-kinetic (I have a Nvidia 1060GTX GPU). The sensor are working individually, and I can add mulitple cpu or gpu sensors as long as I only use one type of sensors in the scene.

To test the issue I first launch an empty gazebo world using:
` roslaunch gazebo_ros empty_world.launch`
Then I spawn the model a modified turtlebot from turtlebot_gazebo and I get the following error:

In the gazebo_ros launch:
```
NODES
  /
    gazebo (gazebo_ros/gzserver)
    gazebo_gui (gazebo_ros/gzclient)<
auto-starting new master
process[master]: started with pid [22009]
ROS_MASTER_URI=http://localhost:11311

setting /run_id to d4f820bc-85c2-11e8-a8d8-9cb6d01b13b7
process[rosout-1]: started with pid [22033]
started core service [/rosout]
process[gazebo-2]: started with pid [22042]
process[gazebo_gui-3]: started with pid [22062]
[ INFO] [1531393248.650194823]: Finished loading Gazebo ROS API Plugin.
[ INFO] [1531393248.650645568]: waitForService: Service [/gazebo/set_physics_properties] has not been advertised, waiting...
[ INFO] [1531393249.665060561, 0.022000000]: waitForService: Service [/gazebo/set_physics_properties] is now available.
[ INFO] [1531393249.692855553, 0.049000000]: Physics dynamic reconfigure ready.
[ INFO] [1531393256.792236413, 6.752000000]: Camera Plugin: Using the 'robotNamespace' param: '/'
[ INFO] [1531393256.796050738, 6.752000000]: Camera Plugin (ns = /)  <tf_prefix_>, set to ""
[ INFO] [1531393256.859807230, 6.752000000]: Laser Plugin: Using the 'robotNamespace' param: '/'
[ INFO] [1531393256.859898936, 6.752000000]: Starting GazeboRosLaser Plugin (ns = /)
[ INFO] [1531393256.861411545, 6.752000000]: GPU Laser Plugin (ns = /) <tf_prefix_>, set to ""
[ INFO] [1531393256.862300586, 6.752000000]: LoadThread function completed
Segmentation fault (core dumped)
[gazebo-2] process has died [pid 22042, exit code 139, cmd /opt/ros/kinetic/lib/gazebo_ros/gzserver -e ode worlds/empty.world __name:=gazebo __log:=/home/antoine/.ros/log/d4f820bc-85c2-11e8-a8d8-9cb6d01b13b7/gazebo-2.log].
log file: /home/antoine/.ros/log/d4f820bc-85c2-11e8-a8d8-9cb6d01b13b7/gazebo-2*.log
 ```
Unfortunately I cannot find the log file....

In the spawn model launch:
```
process[spawn_turtlebot_model-1]: started with pid [22436]
process[mobile_base_nodelet_manager-2]: started with pid [22437]
process[cmd_vel_mux-3]: started with pid [22438]
process[bumper2pointcloud-4]: started with pid [22440]
[ERROR] [1531393257.278663, 6.752000]: Unpause physics service call failed: unable to connect to service: [Errno 104] Connection reset by peer
[spawn_turtlebot_model-1] process has finished cleanly
log file: /home/antoine/.ros/log/d4f820bc-85c2-11e8-a8d8-9cb6d01b13b7/spawn_turtlebot_model-1*.log
```
Logfile:
```
[rosout][INFO] 2018-07-12 13:00:57,127: Spawn status: SpawnModel: Successfully spawned entity
[rospy.internal][WARNING] 2018-07-12 13:00:57,278: Unknown error initiating TCP/IP socket to antoine-Blade:44263 (http://antoine-Blade:38983/): Traceback (most recent call last):
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/impl/tcpros_base.py", line 557, in connect
    self.read_header()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/impl/tcpros_base.py", line 650, in read_header
    self._validate_header(read_ros_handshake_header(sock, self.read_buff, self.protocol.buff_size))
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rosgraph/network.py", line 362, in read_ros_handshake_header
    d = sock.recv(buff_size)
error: [Errno 104] Connection reset by peer

[rospy.internal][WARNING] 2018-07-12 13:00:57,278: Unknown error initiating TCP/IP socket to antoine-Blade:44263 (rosrpc://antoine-Blade:44263): Traceback (most recent call last):
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/impl/tcpros_base.py", line 557, in connect
    self.read_header()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/impl/tcpros_base.py", line 650, in read_header
    self._validate_header(read_ros_handshake_header(sock, self.read_buff, self.protocol.buff_size))
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rosgraph/network.py", line 362, in read_ros_handshake_header
    d = sock.recv(buff_size)
error: [Errno 104] Connection reset by peer

[rospy.internal][INFO] 2018-07-12 13:00:57,278: topic[/clock] removing connection to http://antoine-Blade:38983/
[rosout][ERROR] 2018-07-12 13:00:57,278: Unpause physics service call failed: unable to connect to service: [Errno 104] Connection reset by peer
[rospy.core][INFO] 2018-07-12 13:00:57,279: signal_shutdown [atexit]
[rospy.internal][INFO] 2018-07-12 13:00:57,281: topic[/rosout] removing connection to /rosout
[rospy.impl.masterslave][INFO] 2018-07-12 13:00:57,281: atexit
```

I need this solution because I have to label human being laser hits and the GPU sensors doesn't support laser retro value while the CPU sensor does not sense the "skin" of the gazebo actors. With the 2 sensors a simple point to point subtraction allows to find the laser rays that touches humans. A Depth camera could be used but it does not provide out of the box noise and raytracing.

### PR DESCRIPTION
**Pull request by kev-the-dev**
There are 4 plugins in gazebo_plugins which attach to a ray or gpu_ray sensor and are mostly duplicate code which is confusing for users and is difficult to maintain (changes in one need to be copied to the other):
* gazebo_ros_laser: Attaches to a ray sensor and publishes a sensor_msgs/LaserScan
* gazebo_ros_gpu_laser: Exact duplicate of gazebo_ros_laser but attaches to a gpu_ray
* gazebo_ros_block_laser: Attaches to a ray sensor and publishes a sensor_msgs/PointCloud
* gazebo_ros_range: Attaches to a ray sensor and publishes a sensor_msgs/Range (this one is not affect by this PR)

This PR creates a new plugin gazebo_ros_ray_sensor which can be attached to _either_ a ray or gpu_ray sensor and can publish LaserScan, PointCloud, or PointCloud2 by changing the sdf parameter ```<outputType>```. The old plugins are changed to print a deprecation warning then load this plugin with the necessary configuration to have the same behavior for backwards compatibility during the deprecation "tick".

Implements #772 
Implements #50
Fix #768

### CODE REVIEW

**Comment by chapulina** [comment]
This looks really nice. I'm just worried about API/ABI breakage in a released distribution. The plugin headers being changed here are being installed, so it is possible that downstream users would be subclassing them. I'm not sure what's the ROS policy on that, but we may need to have a more gentle tick.

**Comment by kev-the-dev** [comment]
I suppose I could just add the new plugin and print deprecation warnings in the others. @j-rivero what do you think about preserving API/ABI in melodic for users potentiality sub classing plugins? This PR keeps the plugin's usage/configuration/output the same, but would break things if anyone subclases.

**Comment by dhood** [comment]
I mentioned to @ironmig the risk of very subtle behaviour changes for anyone that happens to be relying on behaviour of `SetActive(false)` of a sensor (e.g. via a pure gazebo plugin).
It's a very specific use case, but that was what we were interested in for ARIAC for the "sensor blackout" challenge.

These plugins have different behaviour in that context, and merging them into a single plugin will change the behaviour of some of them. I don't think it would affect too many people (probably only ARIAC), but it'd be good to keep that in mind e.g. for the migration notes.

https://bitbucket.org/osrf/ariac/pull-requests/91/add-sensor-blackout-agility-challenge/diff has details about how the different sensor behaviour affected ARIAC. Here's the summary:

- gazebo_ros_range and gazebo_ros_block_laser use sensor callbacks to get the sensor data (not a subscription to a gazebo topic), so sensors using those plugins _do_ respond to `SetActive(false)`.
- gazebo_ros_laser uses a gazebo topic subscription to get the sensor data, and since the [RaySensor in gazebo always publishes if there's a subscriber](https://bitbucket.org/osrf/gazebo/src/25df4fb10d33a9baaccdf4d48f99ca7fd4f26584/gazebo/sensors/RaySensor.cc?at=default&fileviewer=file-view-default#RaySensor.cc-511), sensors using that plugin _do not_ respond to `SetActive(false)`.

In the instance that `SetActive(false)` is respected, the sensor will still turn back on if a ROS subscriber connects after the sensor has been set inactive. For ARIAC we added [this change](https://github.com/ros-simulation/gazebo_ros_pkgs/commit/cd34211e01f5b6093f7cfca41535894eaf15a96f) to prevent that from happening, which you might consider for the 'next gen' version of the plugin.

**Comment by awesomebytes** [comment]
Hello, this looks like a nice cleanup. I'd like to ask if in this format the plugin will allow to have multiple plugin instances publishing to the same ROS topic. I've encountered in the past that Gazebo crashes if I try to do that with the gazebo_ros_gpu_laser plugin (and I didn't really find why that was happening).

**Comment by chapulina** [comment]
@awesomebytes , I modified the demo world as follows:

```diff
diff --git a/gazebo_plugins/worlds/gazebo_ros_ray_sensor_demo.world b/gazebo_plugins/worlds/gazebo_ros_ray_sensor_demo.world
index 26b78bc..0fc0466 100644
--- a/gazebo_plugins/worlds/gazebo_ros_ray_sensor_demo.world
+++ b/gazebo_plugins/worlds/gazebo_ros_ray_sensor_demo.world
@@ -166,6 +166,53 @@
           </plugin>
         </sensor>
 
+        <sensor name="sensor_gpu_ray2" type="gpu_ray">
+          <ray>
+            <scan>
+              <horizontal>
+                <samples>300</samples>
+                <resolution>1.0</resolution>
+                <min_angle>-0.5236</min_angle>
+                <max_angle>0.5236</max_angle>
+              </horizontal>
+            </scan>
+            <range>
+              <min>0.05</min>
+              <max>50.0</max>
+            </range>
+          </ray>
+          <always_on>true</always_on>
+          <update_rate>1.0</update_rate>
+          <plugin name="pc2_gpu" filename="libgazebo_ros_ray_sensor.so">
+            <ros>
+              <namespace>/gpu_ray</namespace>
+              <argument>~/out:=pointcloud2</argument>
+            </ros>
+            <output_type>sensor_msgs/PointCloud2</output_type>
+          </plugin>
+          <plugin name="pc_gpu" filename="libgazebo_ros_ray_sensor.so">
+            <ros>
+              <namespace>/gpu_ray</namespace>
+              <argument>~/out:=pointcloud</argument>
+            </ros>
+            <output_type>sensor_msgs/PointCloud</output_type>
+          </plugin>
+          <plugin name="laserscan_gpu" filename="libgazebo_ros_ray_sensor.so">
+            <ros>
+              <namespace>/gpu_ray</namespace>
+              <argument>~/out:=laserscan</argument>
+            </ros>
+            <output_type>sensor_msgs/LaserScan</output_type>
+          </plugin>
+          <plugin name="range_gpu" filename="libgazebo_ros_ray_sensor.so">
+            <ros>
+              <namespace>/gpu_ray</namespace>
+              <argument>~/out:=range</argument>
+            </ros>
+            <output_type>sensor_msgs/Range</output_type>
+          </plugin>
+        </sensor>
+
       </link>
     </model>
```

All the 4 topics seemed to have the correct number of publishers, i.e.:

```bash
$ ros2 topic info /gpu_ray/pointcloud2
Topic: /gpu_ray/pointcloud2
Publisher count: 2
Subscriber count: 0
```

**Comment by mortenfyhn** [comment]
This hasn't had any activity for some time, but I just tested it on my machine, and it works beautifully. Would be wonderful to get this merged sometime.

File: gazebo_plugins/src/gazebo_ros_ray_sensor.cpp [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+  }
+
+  if (RANGE == this->output_type_)
+  {
+    if(!_sdf->HasElement("radiationType"))
+    {
+      ROS_INFO_NAMED("laser", "missing <radiationType>, defaulting to infared");
+      this->range_radiation_type_ = sensor_msgs::Range::INFRARED;
+    } else if ("ultrasound" == _sdf->Get<std::string>("radiationType")) {
+      this->range_radiation_type_ == sensor_msgs::Range::ULTRASOUND;
```
Comment (mortenfyhn) [inline gazebo_plugins/src/gazebo_ros_ray_sensor.cpp:131]: This should probably be an `=` assignment, not `==`. Ditto two lines below.

**Comment by chapulina** [comment]
This is too intrusive to go into a released distribution. It breaks ABI, API and behaviour. I don't think this will ever go into ROS 1, but [it's there on ROS2](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-Migration:-Ray-sensors).

Closing the PR not to encourage false expectations. If someone is willing to write a similar change that leaves current use cases intact, I recommend opening a new PR for it.

**Comment by mortenfyhn** [comment]
I understand, thank you for clarifying.

**Pull request closed without merging**